returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,251 @@
+"""
+Provides :class:`PostprocessingDataset`.
+"""
+
+from __future__ import annotations
+
+from numpy import ndarray
+from typing import Any, Callable, Dict, Iterator, List, Optional, Tuple, Union
+
+from returnn.datasets.basic import DatasetSeq
+from returnn.tensor import Tensor, TensorDict
+from returnn.tensor.dim import Dim
+from .basic import Dataset, init_dataset
+from .cached2 import CachedDataset2
+from .util.strings import str_to_numpy_array
+
+__all__ = ["PostprocessingDataset"]
+
+
+class PostprocessingDataset(CachedDataset2):
+    """
+    A dataset that allows for generic post-processing of data from another dataset
+    using a function on the segment level and on the level of multiply segments via
+    an iterator.
+
+    This allows integrating various data augmentation techniques like e.g. Mixup, or
+    speed perturbation into the data loading pipeline.
+
+    The integration into the data loading pipeline makes it easy to distribute the
+    data processing work across multiple CPU cores using `MultiProcDataset` and in
+    turn frees the GPU from data preprocessing tasks.
+
+    Both functions can be specified at the same time, and their results will be
+    combined by first applying the segment-level function and then the
+    multiple-segments-level function.
+
+    Example usage:
+        train = {
+            "class": "PostprocessingDataset",
+            "dataset": {
+                "class": "HDFDataset",
+                "files": ["/path/to/data.hdf"],
+            },
+            # at least one of them:
+            "map_seq": map_seq,  # (data: TensorDict) -> TensorDict
+            "map_seq_stream": map_seqs,  # (iter: Iterator[TensorDict]) -> Iterator[TensorDict]
+        }
+    """
+
+    def __init__(
+        self,
+        dataset: Dict[str, Any],
+        map_seq: Optional[
+            Union[
+                Callable[[TensorDict], TensorDict],
+                Callable[[TensorDict], Iterator[TensorDict]],
```

Comment:
I don't understand this. When operating on a single seq, it should not return an iterator?